File: ScorchedServer/ScorchedServer/Models/Player.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using ScorchedServer.Properties;

namespace ScorchedServer.Models
{
  public class Player
  {
    public int id { get; set; }
    public string name { get; set; }
    public Vector position { get; set; }
    public double pos { get; set; } // x position
    public string color { get; set; }
    public double health { get; set; }
    public int score { get; set; }
    public double angle { get; set; }
    public double barrelAngle { get; set; }
    private FireRequest lastFireRequest;

    private IEnumerable<Vector> lastShot;

    public Player()
    {
      angle = Convert.ToInt32(Math.PI / 2);
    }
    
    public Rectangle GetRectangle()
    {
      // TODO : Rectangle should be rotated to match player angle.
      // Collission detection should be changed to line intersection.
      var posBellow = Vector.FromCart(this.position.X, this.position.Y - Settings.Default.tankHeight / 2);
      return Rectangle.FromCenter(posBellow, Settings.Default.tankWidth, Settings.Default.tankHeight);
    }

    public IEnumerable<Vector> Shape
    {
      get
      {
        return basicShape().Select(v => v.Rotate(this.angle, this.position));
      }
    }

    private IEnumerable<Vector> basicShape()
    {
      double dx = Settings.Default.tankWidth / 2;
      double dy = Settings.Default.tankHeight / 2;

      yield return Vector.FromCart(this.position.X + dx - Settings.Default.tankGapWidth, this.position.Y + dy);
      yield return Vector.FromCart(this.position.X - dx + Settings.Default.tankGapWidth, this.position.Y + dy);
      yield return Vector.FromCart(this.position.X - dx, this.position.Y - dy);
      yield return Vector.FromCart(this.position.X + dx, this.position.Y - dy);
    }

    public void shoot(FireRequest fr)
    {
      lastFireRequest = fr;

      lastShot = Shot.Trace(position, Vector.FromPolar(fr.angle, fr.power), Vector.FromCart(0, -0.01));
    }

    internal void clearLastShot()
    {
      lastShot = null;
    }

    internal IEnumerable<object> getLastShotAndPlayerHit(IEnumerable<Player> players)
    {
      if (lastShot == null)
        yield break;

      var tanks = from p in players
                  where p != this
                  select new Tuple<Player, IEnumerable<LineSegment>>(p, LineSegment.ClosedPath(p.Shape));
      var lineSegments = lastShot.Zip<Vector, Vector, LineSegment>(lastShot.Skip(1), LineSegment.FromVectors).Take(1000);

      var arc = new List<Vector>();
      Player other = null;

      foreach (var ls in lineSegments)
      {
        var t = intersect(ls, tanks);

        if (t == null)
        {
          arc.Add(ls.V1);
        }
        else
        {
          arc.Add(t.Item2);
          other = t.Item1;

          break;
        }
      }

      if (other != null)
      {
        yield return new { type = "fire", playerId = id, arc = arc.Select(v => new { x = v.X, y = v.Y }) };

        other.health -= 0.1;
        yield return new { type = "updatePlayer", player = other };
      }
    }

    internal Tuple<Player, Vector> intersect(LineSegment ls1, IEnumerable<Tuple<Player, IEnumerable<LineSegment>>> tanks)
    {
       return (from tank in tanks
               from ls2 in tank.Item2
               let intersection = ls1.Intersection(ls2)
               where intersection != null
               select new Tuple<Player, Vector>(tank.Item1, (Vector)intersection)
              ).FirstOrDefault();
    }

    //internal object getLastShot(IEnumerable<Player> players)
    //{
    //  if (lastShot == null)
    //    return null;

    //  var tanks = from p in players
    //              where p != this
    //              select new { player = p, lines = LineSegment.ClosedPath(p.Shape) };

    //  Func<LineSegment, bool> collision = l1 => tanks.Any(tank => tank.lines.Any(l2 => l1.Intersection(l2) != null));

    //  var lineSegments = lastShot.Zip(lastShot.Skip(1), LineSegment.FromVectors);

    //  var limitedShots = lineSegments.TakeWhile(r => !collision(r)).Select(l => l.V1);

    //  var arc = limitedShots.Take(1000).ToArray();

    //  var o = new { type = "fire", playerId = id, arc = arc.Select(v => new { x = v.X, y = v.Y }) };
    //  return o;
    //}

    //internal IEnumerable<object> getPlayersHitPlusWin(IEnumerable<Player> players)
    //{
    //  if (lastShot == null)
    //    yield break;

    //  var playersPlusTanks = players.Select(p => Tuple.Create(p, p.GetRectangle()));

    //  Func<Rectangle, IEnumerable<Player>> collide = r =>
    //    playersPlusTanks.SelectMany(pt => overlap(r, pt.Item2) ? new[] { pt.Item1 } : Enumerable.Empty<Player>());

    //  var rectangles = lastShot.Zip(lastShot.Skip(1), Rectangle.FromVectors);

    //  var playersHit = rectangles.Take(1000).SelectMany(collide);

    //  foreach (var p in playersHit)
    //  {
    //    p.health -= 0.1;

    //    yield return new { type = "updatePlayer", player = p };
    //  }
    //}

    //internal bool overlap(Rectangle r1, Rectangle r2)
    //{
    //  return (r1.LeftTop.X < r2.RightBottom.X && r1.RightBottom.X > r2.LeftTop.X &&
    //        r1.LeftTop.Y > r2.RightBottom.Y && r1.RightBottom.Y < r2.LeftTop.Y);
    //}
  }

  public static class Linq
  {
    public static IEnumerable<C> Zip<A, B, C>(this IEnumerable<A> xs, IEnumerable<B> ys, Func<A, B, C> f)
    {
      var exs = xs.GetEnumerator();
      var eys = ys.GetEnumerator();

      while (exs.MoveNext() && eys.MoveNext())
        yield return f(exs.Current, eys.Current);
    }
  }

  public class Tuple<A, B>
  {
    public A Item1;
    public B Item2;

    public Tuple(A a, B b)
    {
      Item1 = a;
      Item2 = b;
    }

    public static Tuple<C, D> Create<C, D>(C x, D y)
    {
      return new Tuple<C, D>(x, y);
    }
  }
}

File: src/ScorchedServer/play/player.js
function Player(options) {
  this.id = options.id;
  this.name = options.name;
  this.health = options.health;
  this.score = options.score;
  this.angle = options.angle;
  this.barrelAngle = options.barrelAngle;
  this.color = options.color;
  this.shape = options.shape;
  this.pos = options.pos;
  this.posy = options.posy;

  
}

Player.prototype =
{
  computeVectors : function ()
  {
    var tank = {};
    tank.center = Vector.fromCart(this.pos, this.posy)
    tank.rect = Rectangle.fromCenter(tank.center, config.tankWidth, config.tankHeight);
    tank.body = {};
    tank.body.leftTop = tank.rect.leftTop.add(config.tankGapWidth, 0);
    tank.body.rightTop = tank.rect.rightTop.add(-config.tankGapWidth, 0);
    tank.body.leftBottom = tank.rect.leftBottom;
    tank.body.rightBottom = tank.rect.rightBottom;
    tank.barrel = {};
    tank.barrel.start = tank.center;
    tank.barrel.end = tank.center.add(config.barrelLength, 0).rotate(this.barrelAngle - this.angle, tank.barrel.start);
    tank.turret = {};
    tank.turret.center = Vector.fromCart(tank.center.x, tank.rect.bottom - tank.rect.height + 1);
    tank.turret.radius = 2 * config.tankWidth / 4;
    tank.turret.startAngle = Math.PI - (Math.PI / 3.5);
    tank.turret.endAngle = 2 * Math.PI + (Math.PI / 3.5);
    tank.turret.anticlockwise = true;
    tank.health = {};
    tank.health.y = tank.rect.bottom - config.healthIndicatorBottomMargin
    tank.health.start = Vector.fromCart(tank.rect.left, tank.health.y);
    tank.health.middle = Vector.fromCart(tank.rect.left + config.tankWidth * this.health, tank.health.y);
    tank.health.end = Vector.fromCart(tank.rect.left + config.tankWidth, tank.health.y)
    if (this.id == world.me.id)
      world.me.tank = tank;

    // Rotate tank
    tank.body = vectorSpaceRotate(tank.body, this.angle, tank.center);
    tank.barrel = vectorSpaceRotate(tank.barrel, this.angle, tank.center);
    tank.turret = vectorSpaceRotate(tank.turret, this.angle, tank.center);
    tank.turret.startAngle += this.angle;
    tank.turret.endAngle += this.angle;
    tank.health = vectorSpaceRotate(tank.health, this.angle, tank.center);

    this.tank = tank;
  }
}
